docs/Relayer Fee Market.md

```diff
@@ -0,0 +1,84 @@
+
+P1's relayer is R1, P2's relayer is R2, P3's relayer is R3.
+
+The protocol requires R1, R2, R3 to deliver the message and confirm using delivery proof in specific time T. To give them different priority for delivery according to their prices, we will split the T into three consecutive time slots: **_0~T1 for P1_**,  **_T1~T2 for P2_**, and **_T2~T3 for P3_** respectively. The relayer who is assigned with her own time slot will be rewarded with more percentage, and this reward is for the asked price and the commitment(delivery in time or slash). Relayer with lower price are assigned with earlier time slot.
+
+If the relayer fails to deliver in their allotted time slot, the task goes to the next relayer. The penalty fee **_S(t)_** from the previous relayer also adds up to the gain of the current relayer as a bonus, where **_S(t)_** is a function of the delay **_t_**.
+
+### Detailed Steps in Implementation
+
+1. A relayer first registers and deposits some funds (tokens) as the collateral if they are willing to participate.
+    1. enroll_and_lock_collateral dispatch call; relayer enrollment mapping <id, collateral_amount>
+    2. cancel_enrollment() dispatch call (and return collateral, require checkin existing orders)
+2. ***Ask***, asks(price) dispatch call,  updates ask list storage (sorted) 
+    1. Internal call for querying current price(return **_P1_**, **_P2_**, **_P3_** and related relayers)
+    2. RPC
+    3. Update, Cancel
+3. Internal call for bid and execution, some user pay **_P3_** for sending a cross-chain message.
+    1. Create a new order(with current time), in the order lifecycle, relayer cannot cancel enrollment and take back collateral.
+    2. **_P3_** is locked in the Order(contract). 
+    3. Managing orders storage
+4. Internal call for finishing an order or slash an order after message delivery.
+    1. Finish order if this message is delivered by required relayers in the order, and pay **_P3_**.
+    2. Slash order if the message is delivered by other relayers, pay **_P3 + S(t)_** to relayer who delivered message, and slash S(t) from required relayers(of **_P1, P2, P3_**) collateral.
+    3. How to prove this message is delivered by specific relayer?
+        1. When some relayer deliver message, it will attach an extra data describing the source account of relayer.
+        2. The source account will be included in the MessageDelivery event in target chain.
+        3. Thus source account will be included in the Message Delivery Proof.
+    4. Can other relayer help claim with the message delivery proof? And will the protocol incentive this behavior?
+5. If the collateral of any relayer is lower than required collateral threshold (e.g. slashed in order), the enrolment of this relayer will become inactive(will be removed from the ***ask*** list, and not able to put new ***ask***).
+6. Distribution of system revenue
```

Comment:
Point 6's question has been answered in point 7. Maybe we should delete it to avoid make readers confuse.